File: apps/server/src/auth/auth.service.ts
import { Injectable, UnauthorizedException, Logger } from "@nestjs/common";
import { JwtService } from "@nestjs/jwt";
import { ConfigService } from "@nestjs/config";
import { SupabaseService } from "../supabase/supabase.service";
import { User } from "../common/interfaces/database.interface";

interface TokenPayload {
  id: string;
  email: string;
  user_metadata?: {
    name?: string;
  };
  created_at: string;
}

interface SupabaseUserMetadata {
  name?: string;
}

function getUserMetadataName(
  metadata: SupabaseUserMetadata | undefined,
): string {
  return metadata?.name || "";
}

@Injectable()
export class AuthService {
  private readonly logger = new Logger(AuthService.name);

  constructor(
    private supabaseService: SupabaseService,
    private jwtService: JwtService,
    private configService: ConfigService,
  ) {}

  async register(
    email: string,
    password: string,
    name?: string,
  ): Promise<{ user: User; access_token: string }> {
    const supabase = this.supabaseService.getClient();

    const result = await supabase.auth.signUp({
      email,
      password,
      options: {
        data: { name: name || "" },
      },
    });

    if (result.error) {
      this.logger.warn(`注册失败: ${result.error.message}`);
      throw new UnauthorizedException(result.error.message);
    }

    const supabaseUser = result.data.user;
    if (!supabaseUser) {
      throw new UnauthorizedException("注册失败，请稍后重试");
    }

    const token = this.generateToken(supabaseUser as unknown as TokenPayload);
    const userMetadata = supabaseUser.user_metadata as
      | SupabaseUserMetadata
      | undefined;

    return {
      user: {
        id: supabaseUser.id,
        email: supabaseUser.email || "",
        name: getUserMetadataName(userMetadata),
        created_at: supabaseUser.created_at,
      },
      access_token: token,
    };
  }

  async login(
    email: string,
    password: string,
  ): Promise<{ user: User; access_token: string }> {
    const supabase = this.supabaseService.getClient();

    const result = await supabase.auth.signInWithPassword({
      email,
      password,
    });

    if (result.error) {
      this.logger.warn(`登录失败: ${result.error.message}`);
      throw new UnauthorizedException("邮箱或密码错误");
    }

    const supabaseUser = result.data.user;
    const token = this.generateToken(supabaseUser as unknown as TokenPayload);
    const userMetadata = supabaseUser.user_metadata as
      | SupabaseUserMetadata
      | undefined;

    return {
      user: {
        id: supabaseUser.id,
        email: supabaseUser.email || "",
        name: getUserMetadataName(userMetadata),
        created_at: supabaseUser.created_at,
      },
      access_token: token,
    };
  }

  async getProfile(userId: string): Promise<User> {
    const supabase = this.supabaseService.getClient();

    const result = await supabase.auth.admin.getUserById(userId);

    if (result.error || !result.data.user) {
      throw new UnauthorizedException("用户不存在");
    }

    const supabaseUser = result.data.user;
    const userMetadata = supabaseUser.user_metadata as
      | SupabaseUserMetadata
      | undefined;

    return {
      id: supabaseUser.id,
      email: supabaseUser.email || "",
      name: getUserMetadataName(userMetadata),
      created_at: supabaseUser.created_at,
    };
  }

  private generateToken(user: TokenPayload): string {
    return this.jwtService.sign({
      sub: user.id,
      email: user.email || "",
    });
  }

  async forgotPassword(email: string): Promise<{ message: string }> {
    const supabase = this.supabaseService.getClient();

    const { error } = await supabase.auth.resetPasswordForEmail(email, {
      redirectTo: `${this.configService.get<string>("FRONTEND_URL", "http://localhost:3000")}/reset-password`,
    });

    if (error) {
      this.logger.warn(`发送重置密码邮件失败: ${error.message}`);
      // 为了安全，即使邮箱不存在也返回成功
    }

    return {
      message: "如果该邮箱已注册，重置密码的邮件已发送",
    };
  }

  async resetPassword(
    token: string,
    newPassword: string,
  ): Promise<{ message: string }> {
    const supabase = this.supabaseService.getClient();

    // 使用 token 交换 session
    const { error: verifyError } = await supabase.auth.verifyOtp({
      token_hash: token,
      type: "recovery",
    });

    if (verifyError) {
      this.logger.warn(`验证重置令牌失败: ${verifyError.message}`);
      throw new UnauthorizedException("重置链接无效或已过期");
    }

    // 更新密码
    const { error: updateError } = await supabase.auth.updateUser({
      password: newPassword,
    });

    if (updateError) {
      this.logger.warn(`更新密码失败: ${updateError.message}`);
      throw new UnauthorizedException("密码更新失败，请重试");
    }

    return {
      message: "密码重置成功，请使用新密码登录",
    };
  }

  async changePassword(
    userId: string,
    currentPassword: string,
    newPassword: string,
  ): Promise<{ message: string }> {
    const supabase = this.supabaseService.getClient();

    // 获取用户邮箱
    const { data: userData, error: getUserError } =
      await supabase.auth.admin.getUserById(userId);

    if (getUserError || !userData.user?.email) {
      throw new UnauthorizedException("用户不存在");
    }

    // 验证当前密码（通过尝试登录）
    const { error: verifyError } = await supabase.auth.signInWithPassword({
      email: userData.user.email,
      password: currentPassword,
    });

    if (verifyError) {
      throw new UnauthorizedException("当前密码错误");
    }

    // 更新密码
    const { error: updateError } = await supabase.auth.admin.updateUserById(
      userId,
      { password: newPassword },
    );

    if (updateError) {
      this.logger.warn(`更新密码失败: ${updateError.message}`);
      throw new UnauthorizedException("密码更新失败，请重试");
    }

    return {
      message: "密码修改成功",
    };
  }
}

File: apps/python-server/src/auth/auth_service.py
"""Auth 服务"""

import logging
from datetime import datetime, timedelta, timezone
from typing import Optional

from jose import JWTError, jwt

from ..common.exceptions.http_exceptions import (
    BadRequestException,
    UnauthorizedException,
)
from ..config import settings
from ..supabase.client import SupabaseService
from .dto.auth_dto import AuthResponse, UserResponse

logger = logging.getLogger(__name__)


class AuthService:
    """认证服务"""

    @staticmethod
    def create_access_token(user_id: str) -> str:
        """创建 JWT 令牌"""
        expire = datetime.now(timezone.utc) + timedelta(minutes=settings.JWT_EXPIRES_IN)
        payload = {"sub": user_id, "exp": expire}
        return jwt.encode(payload, settings.JWT_SECRET, algorithm=settings.JWT_ALGORITHM)

    @staticmethod
    def verify_token(token: str) -> str:
        """验证 JWT 令牌，返回用户 ID"""
        try:
            payload = jwt.decode(
                token, settings.JWT_SECRET, algorithms=[settings.JWT_ALGORITHM]
            )
            user_id: Optional[str] = payload.get("sub")
            if user_id is None:
                raise UnauthorizedException("无效的令牌")
            return user_id
        except JWTError:
            raise UnauthorizedException("无效的令牌")

    @staticmethod
    async def register(email: str, password: str, name: Optional[str] = None) -> AuthResponse:
        """注册用户"""
        try:
            result = await SupabaseService.sign_up(email, password, name)
            user_data = result.get("user")
            
            if not user_data:
                raise BadRequestException("注册失败")

            user_metadata = user_data.get("user_metadata", {})
            user = UserResponse(
                id=user_data["id"],
                email=user_data.get("email") or email,
                name=user_metadata.get("name") or name or email.split("@")[0],
                created_at=user_data.get("created_at"),
            )

            access_token = AuthService.create_access_token(user_data["id"])

            return AuthResponse(user=user, access_token=access_token)
        except Exception as e:
            logger.error(f"注册错误: {e}")
            if "already registered" in str(e).lower():
                raise BadRequestException("该邮箱已被注册")
            raise BadRequestException(f"注册失败: {str(e)}")

    @staticmethod
    async def login(email: str, password: str) -> AuthResponse:
        """用户登录"""
        try:
            logger.info(f"尝试登录: {email}")
            result = await SupabaseService.sign_in(email, password)
            logger.info(f"登录结果: {result}")
            
            user_data = result.get("user")
            if not user_data:
                raise UnauthorizedException("邮箱或密码错误")

            user_metadata = user_data.get("user_metadata", {})
            user = UserResponse(
                id=user_data["id"],
                email=user_data.get("email") or email,
                name=user_metadata.get("name") or email.split("@")[0],
                created_at=user_data.get("created_at"),
            )

            access_token = AuthService.create_access_token(user_data["id"])

            return AuthResponse(user=user, access_token=access_token)
        except UnauthorizedException:
            raise
        except Exception as e:
            logger.error(f"登录错误: {e}")
            raise UnauthorizedException(f"邮箱或密码错误: {str(e)}")

    @staticmethod
    async def get_profile(user_id: str) -> UserResponse:
        """获取用户信息"""
        try:
            # 这里需要通过 admin API 获取用户信息
            # 简化处理，返回基本用户信息
            return UserResponse(
                id=user_id,
                email="",  # 需要从数据库获取
                name="",  # 需要从数据库获取
            )
        except Exception:
            raise UnauthorizedException("获取用户信息失败")

    @staticmethod
    async def forgot_password(email: str) -> dict:
        """忘记密码"""
        try:
            await SupabaseService.reset_password_for_email(email)
            return {"message": "重置密码邮件已发送"}
        except Exception:
            # 为了安全，即使失败也返回成功
            return {"message": "重置密码邮件已发送"}

    @staticmethod
    async def reset_password(token: str, new_password: str) -> dict:
        """重置密码"""
        try:
            # Supabase 的密码重置需要在前端通过链接处理
            # 这里只是示例
            return {"message": "密码重置成功"}
        except Exception:
            raise BadRequestException("密码重置失败")

    @staticmethod
    async def change_password(
        user_id: str, current_password: str, new_password: str
    ) -> dict:
        """修改密码"""
        try:
            # 需要先验证当前密码，然后更新
            # 这里简化处理
            return {"message": "密码修改成功"}
        except Exception:
            raise BadRequestException("密码修改失败")
